README.md
[![Build Status](https://travis-ci.org/IBM/node-log4js-qradar-syslog-appender.svg?branch=master)](https://travis-ci.org/IBM/node-log4js-qradar-syslog-appender)

# IBM Bluemix DevOps Services - node-log4js-syslog-appender

This module is a qradar syslog appender for node-log4js.

Link to [IBM Cloud DevOps Toolchains](https://cloud.ibm.com/devops).

This is one of hundreds of [IBM Open Source projects at GitHub](http://ibm.github.io).

# License

[The MIT License (MIT)](LICENSE.txt)

# Contributing

Contributions are welcome via Pull Requests. Please submit your very first Pull Request against the [Developer's Certificate of Origin](DCO.txt), adding a line like the following to the end of the file... using your name and email address of course!

Note that all contributions *must* be submitted through pull requests and have to pass the Travis Status Checks in order for code to be merged into master.

Signed-off-by: John Doe <john.doe@example.org>

# Usage

## Pre-requisites
You must be using [log4js-node](https://github.com/nomiddlename/log4js-node) 3.x or higher, and must call `log4js.configure('/path/to/log4js.json')`
somewhere in your code.


## To upgrade
- `npm i log4js-qradar-syslog-appender@latest --save`

## To install
`npm i log4js-qradar-syslog-appender --save`
- Set the following environment variable to true in order to enable the appender: `export log4js_syslog_appender_enabled=true`
- The default behavior is all log messages will be send to syslog, you can override this behavior by
specifying which loggers' log messages to send via the comma separated list env var `export log4js_syslog_appender_whitelist=audit-logs`
- *For local deveopment only*: Add the following appender to your log4js.json file (note this is the minimal valid configuration):
```
{
        "type": "log4js-qradar-syslog-appender",
        "options": {
                "host": "syslog.prd.ccs.ibmcloud.com",
                "port": "6514",
                "product": "otc-api"
          }
}
```
- For production environments (and in source), only push the following in the log4js.json file:
```
{
        "type": "log4js-qradar-syslog-appender",
        "options": {}
}
```
- Set the following env vars (in pipeline - values depending on your setup/app):
```
export log4js_syslog_appender_enabled=true
export log4js_syslog_appender_whitelist=audit-logs,audit-logs-v2
export log4js_syslog_appender_host=syslog.prd.ccs.ibmcloud.com
export log4js_syslog_appender_port=6514
export log4js_syslog_appender_product=otc-api
export log4js_syslog_appender_url=devops.ng.bluemix.net
```

## Use with default syslog

You can use this appender with any default UDP syslog in unencrypted mode.  The environment setup is very similar to above:

```
export log4js_syslog_appender_enabled=true
export log4js_syslog_appender_useUdpSyslog=true
export log4js_syslog_appender_whitelist=audit-logs,audit-logs-v2
export log4js_syslog_appender_host=localhost
export log4js_syslog_appender_port=514
export log4js_syslog_appender_product=otc-api
export log4js_syslog_appender_url=devops.ng.bluemix.net
```


# Setting Certificates
There are two ways of setting the certs, either through a path (meaning you have to check the certs into source control - so not ideal) or preferably, by setting the base64 encoded values as env vars.

## Option 1: Checking them into source control, then specifying the path to them
```
export log4js_syslog_appender_certificatePath=keys/IDS-crt.pem
export log4js_syslog_appender_privateKeyPath=keys/IDS-key.pem
export log4js_syslog_appender_caPath=keys/ca.pem
```

## Option 2: A more secure way is actually setting the cert itself as env vars.
Note: To shorten the length, we use the base64 encoded values of the certs.
```
export log4js_syslog_appender_certificateBase64=zeaalkjsdfkalsdjfkrlasdjflkasjdlfkjsdfKLJFLSKDJF9f34...
export log4js_syslog_appender_privateKeyBase64=pop4545FDSFalkjsdfrkalsdjfklasdjflkasjdlfkjsdfKLJFLSKDJF9f34...
export log4js_syslog_appender_caBase64=ee3rr435F43alkjsdfkalsdjfklasdjflkasjdlfkjsdfKLJFLSKDJF9f34...
```

## Allow connections to servers with self signed certs.  By default, these connections will fail.
```
export log4js_syslog_appender_rejectUnauthorized=false
```


CHANGELOG.md
# Change Log

## [2.0.0](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/2.0.0) (2019-06-22)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v1.0.0...2.0.0)

**Merged pull requests:**

- Dev dependency update [\#23](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/23) ([hermansb](https://github.com/hermansb))
- Update README.md [\#22](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/22) ([hermansb](https://github.com/hermansb))
- Allow log4js 4 peer dependency [\#20](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/20) ([jcayne-ibm](https://github.com/jcayne-ibm))

## [v1.0.0](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v1.0.0) (2018-07-24)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.4.0...v1.0.0)

**Merged pull requests:**

- Update log4js dependency to v3 [\#19](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/19) ([simonkotwicz](https://github.com/simonkotwicz))
- Use github-changelog-generator to auto-generate a CHANGELOG.md file [\#18](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/18) ([hermansb](https://github.com/hermansb))

## [v0.4.0](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.4.0) (2016-12-14)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.3.3...v0.4.0)

**Closed issues:**

- May need to support TCP without mutual auth [\#11](https://github.com/IBM/node-log4js-qradar-syslog-appender/issues/11)
- Need to set destination IP \("dst" attribute\) in event payloads [\#10](https://github.com/IBM/node-log4js-qradar-syslog-appender/issues/10)

**Merged pull requests:**

- Increment version in prep for npm publish [\#17](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/17) ([hermansb](https://github.com/hermansb))
- Function to call is configure rather than config [\#16](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/16) ([hermansb](https://github.com/hermansb))
- Minor fixes [\#15](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/15) ([simonkotwicz](https://github.com/simonkotwicz))
- Support tcp without mutual auth [\#14](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/14) ([hermansb](https://github.com/hermansb))
- add shutdown function [\#9](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/9) ([simonkotwicz](https://github.com/simonkotwicz))
- Update readme to use common log4js\_syslog\_appender\_url [\#8](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/8) ([tcschmid](https://github.com/tcschmid))
- Add build status badge. [\#7](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/7) ([cttttt](https://github.com/cttttt))
- Remove 'fs' dependency [\#6](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/6) ([pdehaan](https://github.com/pdehaan))

## [v0.3.3](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.3.3) (2016-08-22)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.3.2...v0.3.3)

**Closed issues:**

- Strip off "https://" from the front of URLs [\#3](https://github.com/IBM/node-log4js-qradar-syslog-appender/issues/3)

**Merged pull requests:**

- Update release version [\#5](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/5) ([hermansb](https://github.com/hermansb))

## [v0.3.2](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.3.2) (2016-08-22)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.3.1...v0.3.2)

**Merged pull requests:**

- Add support for a url param, strip out preceding https:// [\#4](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/4) ([hermansb](https://github.com/hermansb))

## [v0.3.1](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.3.1) (2016-07-21)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.3.0...v0.3.1)

**Merged pull requests:**

- Support standard syslog UDP/514 [\#2](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/2) ([paulweb515](https://github.com/paulweb515))

## [v0.3.0](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.3.0) (2016-07-06)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.2.2...v0.3.0)

**Merged pull requests:**

- Update readme + increment version [\#1](https://github.com/IBM/node-log4js-qradar-syslog-appender/pull/1) ([hermansb](https://github.com/hermansb))

## [v0.2.2](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.2.2) (2016-07-06)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.2.1...v0.2.2)

## [v0.2.1](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.2.1) (2016-07-01)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.2.0...v0.2.1)

## [v0.2.0](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.2.0) (2016-06-04)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.1.9...v0.2.0)

## [v0.1.9](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.1.9) (2016-06-04)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.1.8...v0.1.9)

## [v0.1.8](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.1.8) (2016-05-30)
[Full Changelog](https://github.com/IBM/node-log4js-qradar-syslog-appender/compare/v0.1.7...v0.1.8)

## [v0.1.7](https://github.com/IBM/node-log4js-qradar-syslog-appender/tree/v0.1.7) (2016-05-18)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

tests/system/basicTcpTest.js
/**
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2016, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 */
 'use strict';

var test = require('tape'),
    log4js = require('log4js'),
    net = require('net');

test('Test message received by tcp server', function(t) {
    t.plan(1);

    var server = net.createServer(function(socket) {
        // 'connection' listener
        console.log('client connected');
        socket.on('end', function() {
            console.log('client disconnected');
        });
        socket.write('hello\r\n');
        socket.pipe(socket);

        socket.on('data', function(data) {
            console.log('received data: ' + data);
            t.ok(data, 'did the message get received over tcp?');
            t.end();
            process.exit(0);
        });

    });
    server.on('error', function(err) {
        throw err;
    });
    server.listen(1514, function(err) {
        if (err) {
            console.log('Error setting up syslog server: ' + JSON.stringify(err, null, 2));
            throw err;
        }
        console.log('server bound on port 1514');

        process.env.log4js_syslog_appender_enabled = 'true';

        log4js.configure({ 
            appenders: {
                qradar: {
                    type: 'log4js-qradar-syslog-appender',
                    options: {
                        host: 'localhost',
                        port: '1514',
                        product: 'basic-tcp-test'
                    }
                }
            },
            categories: { default: { appenders: ['qradar'], level: 'debug' } }
        });

        var logger = log4js.getLogger();
        logger.info('hai');
    });

});


tests/system/basicUdpTest.js
/**
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2016, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 */
 'use strict';

var test = require('tape'),
    log4js = require('log4js'),
    dgram = require('dgram');

test('Test message received by udp server', function(t) {
    t.plan(1);

    var server = dgram.createSocket('udp4');

    server.on('error', function(err) {
      console.log(`server error:\n${err.stack}`);
      server.close();
    });

    server.on('message', function(msg /*, rinfo*/) {
        console.log('received message: ' + msg.toString());
      t.ok(msg, 'did the message get received over udp?');
      t.end();
      process.exit(0);
    });

    server.on('listening', function() {
      console.log('server listening on port 1514');
    });

    server.bind(1514, function(err) {
        if (err) {
            console.log('Error setting up syslog server: ' + JSON.stringify(err, null, 2));
            throw err;
        }

        process.env.log4js_syslog_appender_enabled = 'true';

        log4js.configure({ 
            appenders: {
                qradar: {
                    type: 'log4js-qradar-syslog-appender',
                    options: {
                        host: 'localhost',
                        port: '1514',
                        useUdpSyslog: true,
                        product: 'basic-udp-test'
                    }
                }
            },
            categories: { default: { appenders: ['qradar'], level: 'debug' } }
        });
        var logger = log4js.getLogger('');
        logger.info('hai');

    });

});


tests/unit/testBase64Decode.js
/**
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2016. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 */
 'use strict';

var test = require('tape'),
    base64Decoder = require('../../base64-decode');

test('Test decoded of known encoded text', function(t) {
    var decodedText = base64Decoder('SGVsbG9Xb3JsZDEK');
    t.equal(decodedText, 'HelloWorld1\n', 'was the text decoded properly?');
    t.end();
});

syslog-connection-singleton.js
/**
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2016. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 */
'use strict';

// internal constructor only
var singleton = function() {
    return {
        connection : null,
        droppedMessages: 0,
        circuitBreak: false,
        MAX_TRIES: 3,
        CIRCUIT_BREAK_MINS: 10
    };
};

// returns the current instance if it exists
// otherwise creates a new one.
singleton.getInstance = function() {
    if (!this.instance) {
        this.instance = new singleton();
    }
    return this.instance;
};

// only expose a function that returns the current instance if it exists
// otherwise creates a new one.
module.exports = singleton.getInstance();

log4js-qradar-syslog-appender.js
/**
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2016, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 */
'use strict';

const syslogConnectionSingleton = require('./syslog-connection-singleton'),
    base64Decode = require('./base64-decode'),
    tls = require('tls'),
    fs = require('fs'),
    util = require('util'),
    os = require('os'),
    dgram = require('dgram'),
    net = require('net'),
    log4js = require('log4js');

const logger = log4js.getLogger('log4js-qradar-syslog-appender');

module.exports = {
    appender: appender,
    configure: configure,
    shutdown: shutdown
};

function retryLogic(retryFunction, tries) {
    // we are in circuit break mode. There is something wrong with the qradar connection. We won't try to
    // send any log messages to qradar until the circuit is connected again.
    if (syslogConnectionSingleton.circuitBreak) {
        syslogConnectionSingleton.droppedMessages++;
        return;
    }
    
    // initialize (or increment if already initialized) tries
    if (!tries) {
        tries = 1;
    }
    else {
        tries++;
    }

    if (tries > syslogConnectionSingleton.MAX_TRIES) {
        syslogConnectionSingleton.circuitBreak = true;
        internalLog('Tried sending a message ' + syslogConnectionSingleton.MAX_TRIES + 
            ' times but the client was not connected. ' + 
            'Initiating circuit breaker protocol. ' + 
            'For the next ' + syslogConnectionSingleton.CIRCUIT_BREAK_MINS + 
            ' mins, we will not attempt to send any messages to QRadar.');
        // circuit breaker logic - if detected bad connection, stop trying
        // to send log messages to qradar for syslogConnectionSingleton.CIRCUIT_BREAK_MINS.

        syslogConnectionSingleton.droppedMessages++;
        setTimeout(connectCircuit.bind(this), 
            syslogConnectionSingleton.CIRCUIT_BREAK_MINS * 60 * 1000);
        return;
    }
    setTimeout(retryFunction.bind(this, tries), 100);
    return;
}

function connectCircuit() {
    internalLog('Re-connecting the circuit. So far we have dropped ' + 
        syslogConnectionSingleton.droppedMessages + ' messages.');
    syslogConnectionSingleton.circuitBreak = false;
}

function readBase64StringOrFile(base64, file, callback) {
    if (base64) {
        callback(null, base64Decode(base64));
    } else {
        fs.readFile(file, {'encoding': 'utf8'}, callback);
    }
}

function loggingFunction(options, log, tries) {

    // internal log message. Do not cause infinite loop by attempting to syslog server.
    if (log.categoryName === 'log4js-qradar-syslog-appender') return;

    if (syslogConnectionSingleton.shutdown) {
        return;
    }
    // we are in circuit break mode. There is something wrong with the qradar connection. We won't try to
    // send any log messages to qradar until the circuit is connected again.
    if (syslogConnectionSingleton.circuitBreak) {
        syslogConnectionSingleton.droppedMessages++;
        return;
    }

    if (!syslogConnectionSingleton.connection && !syslogConnectionSingleton.connecting) {
        syslogConnectionSingleton.connecting = true;
        // udp
        if (options.useUdpSyslog) {
            attemptUdpConnection(options, log, tries);
        }
        else { // tcp
            var boundFunction = attemptTcpConnection.bind(this, log, tries);
            if (mutualAuthConfigured(options)) {
                configureMutualAuth(options, boundFunction);
            }
            else {
                boundFunction(options);
            }
        }
    }
    else {
        logMessage(log, options, tries);
    }
}

function attemptUdpConnection(options, log, tries) {
    var client = dgram.createSocket('udp4');
    syslogConnectionSingleton.connection = {
        write: function (msg) {
            client.send(msg, 0, msg.length, options.port, options.host, function (err) {
                if (err && err !== 0) {
                    cleanupConnection(err, 'error');
                    retryLogic(loggingFunction.bind(this, options, log), tries);
                }
            });
        },
        destroy: function() {
            client.close();
        }
    };
    client.on('error', function(err) {
        cleanupConnection(err, 'error');
        retryLogic(loggingFunction.bind(this, options, log), tries);
    });
    syslogConnectionSingleton.connecting = false;
    logMessage(log, options, tries);
}

function mutualAuthConfigured(options) {
    return ( (options.certificateBase64 || options.certificatePath) &&
        (options.privateKeyBase64 || options.privateKeyPath) &&
        (options.caBase64 || options.caPath) );
}

// set up mutual auth.
function configureMutualAuth(options, callback) {

    readBase64StringOrFile(options.certificateBase64, options.certificatePath, function(err, certificate) {
        if (err) {
            logger.error('Error while loading certificate from path: ' + options.certificatePath + ' Error: ' + JSON.stringify(err, null, 2));
            return;
        }

        options.certificate = certificate;

        readBase64StringOrFile(options.privateKeyBase64, options.privateKeyPath, function(err, key) {
            if (err) {
                logger.error('Error while loading private key from path: ' + options.privateKeyPath + ' Error: ' + JSON.stringify(err, null, 2));
                return;
            }

            options.key = key;

            readBase64StringOrFile(options.caBase64, options.caPath, function(err, caCert) {
                if (err) {
                    logger.error('Error while loading ca key from path: ' + options.caPath + ' Error: ' + JSON.stringify(err, null, 2));
                    return;
                }

                options.caCert = caCert;

                callback(options, true);
            });
        });
    });
}

function attemptTcpConnection(log, tries, options, mutualAuth) {
    var tlsOptions = {
        cert: options.certificate,
        key: options.key,
        ca: options.caCert,
        host: options.host,
        port: options.port,
        passphrase: options.passphrase,
        facility: options.facility,
        tag: options.tag,
        leef: options.leef,
        vendor: options.vendor,
        product: options.product,
        product_version: options.product_version,
        rejectUnauthorized: options.rejectUnauthorized
    };

    var tcpLib;
    
    if (mutualAuth) {
        tcpLib = tls;
    }
    else {
        tcpLib = net;
    }

    syslogConnectionSingleton.connection = tcpLib.connect(tlsOptions, connected.bind(this, log, options, tries));

    syslogConnectionSingleton.connection.setEncoding('utf8');
    syslogConnectionSingleton.connection.on('error', function(err) {
        cleanupConnection(err, 'error');
        retryLogic(loggingFunction.bind(this, options, log), tries);
    });
    syslogConnectionSingleton.connection.on('close', function(err) {
        cleanupConnection(err, 'closed');
        retryLogic(loggingFunction.bind(this, options, log), tries);
    });
    syslogConnectionSingleton.connection.on('end', function(err) {
        cleanupConnection(err, 'ended');
        retryLogic(loggingFunction.bind(this, options, log), tries);
    });
}

function cleanupConnection(err, type) {
    logger.warn('Connection ' + type + '. Error: ' + JSON.stringify(err, null, 2));
    if (syslogConnectionSingleton.connection) {
        syslogConnectionSingleton.connection.destroy();
        syslogConnectionSingleton.connection = null;
    }
    syslogConnectionSingleton.connecting = false;
}

function appender(config) {

    var options = {
        host: process.env.log4js_syslog_appender_host || config.options && config.options.host,
        port: process.env.log4js_syslog_appender_port || config.options && config.options.port,
        useUdpSyslog: process.env.log4js_syslog_appender_useUdpSyslog !== undefined ? process.env.log4js_syslog_appender_useUdpSyslog : config.options && config.options.useUdpSyslog,
        certificatePath: process.env.log4js_syslog_appender_certificatePath || config.options && config.options.certificatePath,
        privateKeyPath: process.env.log4js_syslog_appender_privateKeyPath || config.options && config.options.privateKeyPath,
        passphrase: process.env.log4js_syslog_appender_passphrase || config.options && config.options.passphrase || '',
        caPath: process.env.log4js_syslog_appender_caPath || config.options && config.options.caPath,
        certificateBase64: process.env.log4js_syslog_appender_certificateBase64 || config.options && config.options.certificateBase64,
        privateKeyBase64: process.env.log4js_syslog_appender_privateKeyBase64 || config.options && config.options.privateKeyBase64,
        caBase64: process.env.log4js_syslog_appender_caBase64 || config.options && config.options.caBase64,
        facility: process.env.log4js_syslog_appender_facility || config.options && config.options.facility || '',
        tag: process.env.log4js_syslog_appender_tag || config.options && config.options.tag || '',
        leef: process.env.log4js_syslog_appender_leef || config.options && config.options.leef || '',
        vendor: process.env.log4js_syslog_appender_vendor || config.options && config.options.vendor || '',
        product: process.env.log4js_syslog_appender_product || config.options && config.options.product,
        product_version: process.env.log4js_syslog_appender_product_version || config.options && config.options.product_version || '',
        rejectUnauthorized: process.env.log4js_syslog_appender_rejectUnauthorized !== undefined ? process.env.log4js_syslog_appender_rejectUnauthorized : config.options && config.options.rejectUnauthorized,
        url: process.env.log4js_syslog_appender_url || config.options && config.options.url || process.env.url || os.hostname() || ''
    };

    var stripOut = ['https://', 'http://'];
    for (var i = 0; i < stripOut.length; i++) {
        if (options.url.startsWith(stripOut[i])) {
            options.url = options.url.slice(stripOut[i].length);
        }
    }
    
    // make sure boolean flags work properly with string inputs
    options.useUdpSyslog = parseToBoolean(options.useUdpSyslog); // default is false
    options.rejectUnauthorized = parseToBoolean(options.rejectUnauthorized, true); // default is true

    if (!verifyOptions(options)) {
        return function() {};
    }

    // deep clone of options
    var optionsClone = JSON.parse(JSON.stringify(options));
    delete optionsClone.privateKeyBase64;
    internalLog('Syslog appender is enabled with options: ' + JSON.stringify(optionsClone, null, 2), true);

    syslogConnectionSingleton.shutdown = false;

    return loggingFunction.bind(this, options);
}

function connected(message, options, tries) {
    syslogConnectionSingleton.connecting = false;
    logger.warn('We have (re)connected to QRadar using a secure connection with ' +
        (syslogConnectionSingleton.connection.authorized ? 'a valid ' : 'an INVALID ') +
        'peer certificate. ' + syslogConnectionSingleton.droppedMessages + ' messages have been dropped.');
    logMessage(message, options, tries);
}

function logMessage(log, options, tries) {

    // we are in circuit break mode. There is something wrong with the qradar connection. We won't try to
    // send any log messages to qradar until the circuit is connected again.
    if (syslogConnectionSingleton.circuitBreak) {
        syslogConnectionSingleton.droppedMessages++;
        return;
    }

    // we got disconnected or are still connecting
    // retry later when we are hopefully (re)connected
    if (!syslogConnectionSingleton.connection || syslogConnectionSingleton.connecting) {
        return retryLogic(loggingFunction.bind(this, options, log), tries);
    }

    // if theres a whitelist then only send those messages
    var logWhitelist = process.env.log4js_syslog_appender_whitelist;
    var categoriesToSend = logWhitelist && logWhitelist.split(',');
    if (logWhitelist && categoriesToSend.indexOf(log.categoryName) === -1) return;

    var formattedMessage = formatMessage(log.data.join(' | '), log.level && log.level.levelStr, options);

    syslogConnectionSingleton.connection.write(formattedMessage);
}

function levelToSeverity(levelStr) {
    var levels = [
        'FATAL',
        'ERROR',
        'WARN',
        'INFO',
        'DEBUG',
        'TRACE'
    ];

    return levels.indexOf(levelStr) !== -1 ? levels.indexOf(levelStr) + 1 : 4;
}

function formatMessage(message, levelStr, options) {
    // format as:
    // HEADER STRUCTURED_DATA MESSAGE
    // where HEADER = <PRI>VERSION TIMESTAMP HOSTNAME APP-NAME PROCID MSGID
    // for details see the RFC here http://tools.ietf.org/html/rfc5424# .
    return util.format(
        '<%d>%d %s %s %s %d %s %s %s\n',
        16*8+levelToSeverity(levelStr), // hardcoded facility of local0 which translates to 16 according to RFC. TODO: use passed in facility.
        1,
        new Date().toJSON(),
        options.url,
        options.product,
        process.pid,
        '-', 
        '-',
        message || '{}'
    );
}

function configure(config) {
    if (process.env.log4js_syslog_appender_enabled !== 'true') {
        return function() {};
    } else {
        return appender(config);
    }
}

function parseToBoolean(val, defaultValue) {
    if (defaultValue) { // default is true
        return val !== 'false' && val !== false;
    }
    else { // default is false
        return val === 'true' || val === true;
    }
}

function verifyOptions(options) {
    var requiredOptions = [
        'log4js_syslog_appender_host',
        'log4js_syslog_appender_port',
        'log4js_syslog_appender_product'
    ];
    var valid = true;

    requiredOptions.forEach(function(option) {
        var key = option.substring(option.lastIndexOf('_')+1);
        if (!options[key]) {
            internalLog(key + ' is a required option. It is settable with the ' +
                option + ' environment variable.', true);
            valid = false; // array.forEach is blocking
        }
    });

    
    [ 
        'log4js_syslog_appender_certificate',
        'log4js_syslog_appender_privateKey',
        'log4js_syslog_appender_ca',
    ].forEach(function (option) {
        var key = option.split('_').pop();

        if (!options[key + "Path"] && !options[key + "Base64"] && !options.useUdpSyslog) {
            internalLog('In order to enable ' + 'mutual auth, either ' + key +
                'Path or ' + key +
                'Base64 are required options. It is settable with the ' +
                option + ' environment variable.', true);
        }

        // Deprecated warnings.
        if (options[key + "Path"]) {
            if (options.useUdpSyslog) {
                internalLog('WARNING env var ' + key +
                    'Path will not be used for unencrypted syslog UDP/514.', true);
            } else {
                internalLog('WARNING env var ' + key + 
                    'Path is now deprecated and will be removed in a future' + 
                    ' release. Please switch to ' + key + 'Base64 instead.', true);
            }
        }
        if (options[key + "Base64"] && options.useUdpSyslog) {
            internalLog('WARNING env var ' + key + 
                'Base64 will not be used for unencrypted syslog UDP/514.', true);
        }
    })

    return valid;
}

function shutdown(callback) {
    syslogConnectionSingleton.shutdown = true;
    cleanupConnection('log4js is shutting down', 'shutting down');
    callback();
}

function internalLog(msg, delay) {
    if (delay) { // let log4js configuration to complete before sending a message using it.
        setTimeout(() => {
                logger.info(`${msg}`);
            }, 5000);
    }
    else {
        logger.info(`${msg}`);
    }
}